File: Front/pongJs/Ball.js
class Ball extends MovingRectangle {
    #collide

    constructor(position = { x: 0, y: 0 }, speed = { x: 0, y: 0 }, 
        size = { x: 0, y: 0 }, color = "white") {
        super();
        super.initialize(position, speed, size, color);
        this.#collide = [];
    }

    checkCollision(collider) {
        // Check if upper right corner is touching the collider
        if (collider.getCorners.xl <= this.getCorners.xl &&
            this.getCorners.xl <= collider.getCorners.xh &&
            collider.getCorners.yl <= this.getCorners.yl &&
            this.getCorners.yl <= collider.getCorners.yh) {
            return true;
        }
        // Check if upper left corner is touching the collider
        if (collider.getCorners.xl <= this.getCorners.xh &&
            this.getCorners.xh <= collider.getCorners.xh &&
            collider.getCorners.yl <= this.getCorners.yl &&
            this.getCorners.yl <= collider.getCorners.yh) {
            return true;
        }
        // Check if lower right corner is touching the collider
        if (collider.getCorners.xl <= this.getCorners.xl &&
            this.getCorners.xl <= collider.getCorners.xh &&
            collider.getCorners.yl <= this.getCorners.yh &&
            this.getCorners.yh <= collider.getCorners.yh) {
            return true;
        }
        // Check if lower left corner is touching the collider
        if (collider.getCorners.xl <= this.getCorners.xh &&
            this.getCorners.xh <= collider.getCorners.xh &&
            collider.getCorners.yl <= this.getCorners.yh &&
            this.getCorners.yh <= collider.getCorners.yh) {
            return true;
        }
        return false;
    }

    collisionSpeedSetter() {

    }

    collisionHandler(colider) {
        if (this.checkCollision(colider)) {
            const leftInter = Math.max(this.getCorners.xl, colider.getCorners.xl);
            const rightInter = Math.min(this.getCorners.xh, colider.getCorners.xh);
            const topInter = Math.max(this.getCorners.yl, colider.getCorners.yl);
            const bottomInter = Math.min(this.getCorners.yh, colider.getCorners.yh);
            const adj = Math.abs(leftInter - rightInter) / Math.abs(this.getSpeed.x);
            this.setPosition({
                x: this.getPosition.x - adj * this.getSpeed.x,
                y: this.getPosition.y - adj * this.getSpeed.y
            });
            const x1 = Math.abs(this.getCorners.xl - colider.getCorners.xh);
            const x2 = Math.abs(this.getCorners.xh - colider.getCorners.xl);
            const y1 = Math.abs(this.getCorners.yl - colider.getCorners.yh);
            const y2 = Math.abs(this.getCorners.yh, colider.getCorners.yh);

            if (Math.min(x1, x2) < Math.min(y1, y2)) {
                super.setSpeed({ x: -super.getSpeed.x , y: this.getSpeed.y 
                                                    + 0.5 * colider.getSpeed.y});
            } else {
                super.setSpeed({ y: -super.getSpeed.y });
            }
        }

    }

    updatePosition() {
        if (super.getPosition.x + super.getSize.x >= canvas.width) {
            super.setSpeed({ x: -super.getSpeed.x });
            return 1;
        }
        if (super.getPosition.x < 0) {
            super.setSpeed({ x: -super.getSpeed.x });
            return -1;
        }
        if (super.getPosition.y + super.getSize.y >= canvas.height
            || super.getPosition.y < 0) {
            super.setSpeed({ y: -super.getSpeed.y });
        }
        Object.values(this.#collide).forEach(ele => this.collisionHandler(ele));
        super.setPosition({
            x: super.getPosition.x + super.getSpeed.x,
            y: super.getPosition.y + super.getSpeed.y
        })
        return 0;
    }

    addColider(colider) {
        this.#collide.push(colider);
    }

}

File: python-pong/Ball.py
from MovingRectangle import MovingRectangle

class Ball(MovingRectangle):
	# enclousure is NOT on the JS version of this, it is here to have values 
	# that are in the global canvas.height and canvas.width
	def __init__(self, dictCanvas: dict, name: str,
			  position = {"x" : 0, "y" : 0}, speed = {"x" :0, "y" : 0},
			  size = {"x" : 0, "y" : 0}, enclousure = None):
		self.initialize(dictCanvas = dictCanvas, name = name,
				   position = position,speed = speed, 
				   size = size, enclousure = enclousure)
		self._collide = []
		

	def checkCollision(self, colider: MovingRectangle) -> bool:
		# Check if upper right corner is touching the collider
		if (colider.getCorners()["xl"] <= self.getCorners()["xl"] and 
			self.getCorners()["xl"] <= colider.getCorners()["xh"] and
			colider.getCorners()["yl"] <= self.getCorners()["yl"] and
			self.getCorners()["yl"] <= colider.getCorners()["yh"]):
			return True

		#Check if upper left corner is touching the collider
		if (colider.getCorners()["xl"] <= self.getCorners()["xh"] and 
			self.getCorners()["xh"] <= colider.getCorners()["xh"] and
			colider.getCorners()["yl"] <= self.getCorners()["yl"] and
			self.getCorners()["yl"] <= colider.getCorners()["yh"]):
			return True

		#// Check if lower right corner is touching the collider
		if (colider.getCorners()["xl"] <= self.getCorners()["xl"] and 
			self.getCorners()["xl"] <= colider.getCorners()["xh"] and
			colider.getCorners()["yl"] <= self.getCorners()["yh"] and
			self.getCorners()["yh"] <= colider.getCorners()["yh"]):
			return True
        
		#// Check if lower left corner is touching the collider
		if (colider.getCorners()["xl"] <= self.getCorners()["xh"] and 
			self.getCorners()["xh"] <= colider.getCorners()["xh"] and
			colider.getCorners()["yl"] <= self.getCorners()["yh"] and
			self.getCorners()["yh"] <= colider.getCorners()["yh"]):
			return True
		return False
	
	def collisionHandler(self, colider: MovingRectangle):
		#//Calculate the left and right side of the "inscribed square of the intersection"
		leftInter = max(self.getCorners()["xl"],colider.getCorners()["xl"])
		rightInter = min(self.getCorners()["xh"],colider.getCorners()["xh"])
		# calculate the proportion of the movement that overlaps
		adj = abs(leftInter - rightInter)/abs(self.getSpeed()["x"])
		# undo overlaping
		self.setPosition(x = self.getPosition()["x"] - adj * self.getSpeed()["x"],
				   y = self.getPosition()["y"] - adj * self.getSpeed()["y"])
		#calculate the nearest side and consider that colision from that dir
		x1 = abs(self.getCorners()["xl"] - colider.getCorners()["xh"])
		x2 = abs(self.getCorners()["xh"] - colider.getCorners()["xl"])
		y1 = abs(self.getCorners()["yl"] - colider.getCorners()["yh"])
		y2 = abs(self.getCorners()["yh"] - colider.getCorners()["yl"])
		if min(x1,x2) < min(y1,y2):
			self.setSpeed(x = -self.getSpeed()["x"], 
				  y = self.getSpeed()["y"] + 0.5 * colider.getSpeed()["y"])
		else:
			self.setSpeed(y = -self.getSpeed()["y"])
		
	def updatePosition(self) ->int:
		if (self.getPosition()["x"] + self.getSize()["x"] >= 
	  			self.getEnclousure()["xh"]):
			self.setSpeed(x = -self.getSpeed()["x"])
			return 1
		if (self.getPosition()["x"] < self.getEnclousure()["xl"]):
			self.setSpeed(x = -self.getSpeed()["x"])
			return -1
		if (self.getPosition()["y"] + self.getSize()["y"] > 
	  			self.getEnclousure()["yh"] 
				or self.getPosition()["y"] < self.getEnclousure()["yl"]):
			self.setSpeed(y = -self.getSpeed()["y"])
		for ele in self._collide:
			if (self.checkCollision(ele)):
				self.collisionHandler(ele)
		self.setPosition(x = self.getPosition()["x"] + self.getSpeed()["x"],
				   y = self.getPosition()["y"] + self.getSpeed()["y"])
		return 0
	
	def addColider(self,colider):
		self._collide.append(colider)
